---
title: "04-RNAseq_visual_data_exploration"
author: "Valentin Marteau"
params:
  data: "../data"
  lib:  "../lib"
  results:  "../results"
output:
  html_notebook:
    theme: spacelab
    highlight: textmate
    toc: yes
    number_sections: true
    toc_depth: 3
    toc_float: true
---

```{r, results = "hide"}
# Load required packages
library(conflicted)
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PoiClaClu)
library(variancePartition)
library(gridExtra)
library(cowplot)

conflict_prefer("select", "dplyr")
conflict_prefer("filter", "dplyr")
```

```{r}
# Read data into memory
count_mat <- readRDS(file = file.path(params$results, "produced_data/01-count_mat.rds"))
metadata <- readRDS(file = file.path(params$results, "produced_data/01-metadata.rds"))

# Set ggplot basic theme
theme_set(
  theme(panel.grid = element_blank(),
        text = element_text(size = 26),
        plot.title = element_text(size = 16, hjust = 0.5),
        axis.line = element_line(),
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid.minor = element_line(linetype = "dotted"))
  )
```

# Normalize counts using DESeq2 rlog and vst transformations for exploratory analysis
Have a look at [End-to-end RNA-Seq workflow 2015](https://www.bioconductor.org/help/course-materials/2015/CSAMA2015/lab/rnaseqCSAMA.html)
```{r}
dds <- DESeqDataSetFromMatrix(countData = count_mat,
                              colData = metadata,
                              design = ~condition)

# rlog and vst transformation
dds <- estimateSizeFactors(dds)
rld <- rlog(dds)
vst <- vst(dds, blind = TRUE)
fpm <- fpm(dds)
#fpkm
```

```{r}
plot_layout <- function(df, x_string, y_string, title) {
  ggplot(df, aes_string(x = x_string, y = y_string)) +
    geom_point(size = 0.3) +
    ggtitle(title)
}

# Have a look at different transformations by plotting first against second sample
p <- plot_grid(plot_layout(as_tibble(log2(1 + counts(dds, normalized = TRUE))),"Ctrl_1", "Ctrl_2", "log2"),
               plot_layout(as_tibble(assay(rld)),"Ctrl_1", "Ctrl_2", "rlog"),
               plot_layout(as_tibble(assay(vst)),"Ctrl_1", "Ctrl_2", "vst"),
               plot_layout(as_tibble(log2(fpm)),"Ctrl_1", "Ctrl_2", "fpm"))

ggsave(file.path(params$results, "figures/04-Compare-Transformations.pdf"), plot = p, width = 297, height = 210, units = "mm")
```
Based on the plots above I will use the rld data for subsequent plots.

# Principal Component Analysis
```{r}
# DESeq2 PCA
#plotPCA(rld, intgroup = "condition")
```
## Plot pca using ggplot
```{r}
# Extract pca coordinates
pcaData <- plotPCA(rld, intgroup = c("condition"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"), digits = 1)

p <- ggplot(pcaData, aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = condition),
             size = 8,
             shape = 21) +
  scale_fill_manual(values = c("#A6D854", "#FC8D62", "#80B1D3")) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggtitle("PCA Transcriptome")

ggsave(file.path(params$results, "figures/04-PCA.pdf"), plot = p, width = 297, height = 210, units = "mm")
```

# Sample distances
```{r}
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- rld$label
colnames(sampleDistMatrix) <- rld$condition
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
p <- pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors,
         main = "Euclidean distance between samples (rlog transformed)")

ggsave(file.path(params$results, "figures/04-Sample-distances_Euclidean.pdf"), plot = p, width = 297, height = 210, units = "mm")
```
```{r}
poisd <- PoissonDistance(t(counts(dds)))

samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- rld$label
colnames(samplePoisDistMatrix) <- rld$condition
p <- pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colors,
         main = "Poisson distance (raw data)")

ggsave(file.path(params$results, "figures/04-Sample-distances_Poisson.pdf"), plot = p, width = 297, height = 210, units = "mm")
```

# Pairwise sample correlations
```{r}
# Compute pairwise correlation values
rld_cor <- cor(assay(rld))
p <- pheatmap(rld_cor,
         annotation = select(metadata, condition))

ggsave(file.path(params$results, "figures/04-Pairwise-correlations.pdf"), plot = p, width = 297, height = 210, units = "mm")
```

# Partitioning gene expression variability into different components
Assessing the contribution of variance in expression models at the gene level using the "variancePartition" package
```{r}
# Estimate library size correction scaling factors
dds <- estimateSizeFactors(dds)

# identify genes that pass expression cutoff
isexpr <- rowSums(fpm(dds) > 1) >= 0.5 * ncol(dds)

# compute log2 Fragments Per Million
# Alternatively, fpm(), fpkm(), vst() or rlog() could be used
#quantLog <- log2(fpm(dds)[isexpr, ] + 1)
quantLog <- assay(rlog(dds))[isexpr, ] + 1

# Define formula
form <- ~ (1|condition)

# Run variancePartition analysis
varPart <- fitExtractVarPartModel(quantLog, form, metadata)

# sort variables (i.e. columns) by median fraction of variance explained
vp <- sortCols(varPart)
```
```{r}
# Bar plot of variance fractions for the first 10 genes
plotPercentBars(vp[1:10, ])
```
```{r}
# violin plot of contribution of each variable to total variance
p <- plotVarPart(vp)
ggsave(file.path(params$results, "figures/04-variancePartition.pdf"), plot = p, width = 297, height = 210, units = "mm")
```

```{r}
# sort genes based on variance explained by Individual features
head(varPart[order(varPart$condition, decreasing = TRUE), ])

median(varPart$condition)
median(varPart$Residuals)
```